%!TEX root =  proposal.tex

\subsection{The Team}

% The Project Description must contain a section titled "Research Areas" that must:

% Explicitly state and motivate at least four research areas covered (along with senior personnel with commensurate expertise);
% Describe the targeted distributed applications and systems, and the heterogeneous platforms on which they run; and
% Define relevant notions of scale and describe how scalability will be theoretically and experimentally evaluated for the targeted distributed applications and systems and heterogeneous platforms in (2) with respect to the full hardware/software stack.

The team has a rare combination of skills needed for this proposal.  \defn{We are a tightly coupled team:}  All of the PIs have written multiple papers with other PIs, and we have 19 published papers, each authored by at least three  of the PIs.
%three of the PIs have published papers with at least three other PIs.  
\defn{The proposed work is in our main area of expertise:} Every PI has published in HPC, Systems, Computational Biology and Algorithms.  More specifically:
%One of the PIs is a world leader in GPUs, two of the PIs are world leaders in Computational Biology, two  of the PIs are world leaders in the theory of data structures, and two of the PIs are world leaders in high-performance data structures. All of us have as a main thrust of our research the transfer of theoretical results in data structures to practice, especially in HPC settings. Three of us have particular expertise in implementing sophisticated data structures in multi-node multi-GPUs settings.


\begin{itemize}%[noitemsep,nolistsep]
    \item \textbf{GPU Systems}: PIs Farach-Colton, Owens, and Pandey

    \item \textbf{High-Performance Computing}: PIs Bender, Farach-Colton, Owens, and Pandey

    \item \textbf{Large-scale computational biology} PIs Bender, Farach-Colton, Pandey, and Patro

        \item \textbf{Theory and Algorithms}: PIs Bender, Farach-Colton, Pandey, and Patro
    

\end{itemize}



%\paragraph{Notions of Scale.}
Our proposed work addresses several notions of scale.  First, our work involves scaling up data structures by designing them for GPUs. GPUs are cost-effective and offer massive parallelism, allowing significant speedups compared to CPUs. GPU data structures can help speed up computational-biology applications and quickly analyze large-scale datasets. Without a principled redesign of data structures, additional device RAM and GPU cores will be of diminishing value. Second, our work includes scaling out data structures in distributed memory across multi-node GPUs to quickly process petabyte-scale genomic and metagenomic datasets. This will involve building distributed data structures that can offer low communication volume and low load imbalance. Prior work has demonstrated that data movement and load imbalance is the major bottleneck for achieving high performance in a distributed application. Furthermore, computational biology datasets available today are already terabyte- and petabyte-scale. For example, raw sequencing data from SRA~\cite{kodama2012sequence}, metagenomic data from WA and Rhizo~\cite{hofmeyr2020terabase}, and pangenomic data from the 100,000 Genome Project~\cite{1002021100}.  To quickly process and perform biological analysis on these data, we need to exploit the massive computing in modern GPUs (V100 and A100) and also the distributed computing infrastructure of supercomputers (Perlmutter~\cite{perlmutter} and Summit~\cite{summit}).
